yajsapi/utils/log.ts

```diff
@@ -1,37 +1,303 @@
-    debug: "magenta",
-    warn: "yellow",
-    error: "red"
-}
+  info: "blue",
+  debug: "magenta",
+  warn: "yellow",
+  error: "red",
+};
 winston.addColors(logColors);
 const logger = winston.createLogger(options);
 
+class SummaryLogger {
+  private _wrapped_emitter: Callable<[events.YaEvent], void> | null;
+  public numbers: number;
+
+  // Start time of the computation
+  start_time!: number;
+
+  // Maps received proposal ids to provider ids
+  received_proposals!: { [key: string]: string };
+
+  // Set of confirmed proposal ids
+  confirmed_proposals!: Set<string>;
+
+  // Maps agreement ids to provider names
+  agreement_provider_name!: { [key: string]: string };
+
+  // Set of agreements confirmed by providers
+  confirmed_agreements!: Set<string>;
+
+  // Maps task id to task data
+  task_data!: { [key: string]: any };
+
+  // Maps a provider name to the list of task ids computed by the provider
+  provider_tasks!: { [key: string]: string[] };
+
+  // Map a provider name to the sum of amounts in this provider's invoices
+  provider_cost!: { [key: string]: number };
+
+  // Count how many times a worker failed on a provider
+  provider_failures!: { [key: string]: number };
+
+  // Has computation finished?
+  finished!: boolean;
+
+  error_occurred!: boolean;
+
+  time_waiting_for_proposals;
+
+  constructor(wrapped_emitter: Callable<[events.YaEvent], void> | null = null) {
+    this._wrapped_emitter = wrapped_emitter;
+    this.numbers = 0;
+    this._reset();
+  }
+
+  _reset(): void {
+    this.start_time = dayjs().valueOf() / 1000;
+    this.received_proposals = {};
+    this.confirmed_proposals = new Set();
+    this.agreement_provider_name = {};
+    this.confirmed_agreements = new Set();
+    this.task_data = {};
+    this.provider_tasks = {};
+    this.provider_cost = {};
+    this.provider_failures = {};
+    this.finished = false;
+    this.error_occurred = false;
+    this.time_waiting_for_proposals = dayjs.duration(0);
+  }
+
+  _print_total_cost(): void {
+    if (!this.finished) return;
+    const provider_names = new Set(Object.keys(this.provider_tasks));
+    if (isSuperset(provider_names, new Set(Object.keys(this.provider_cost)))) {
+      const total_cost = Object.values(this.provider_cost).reduce(
+        (item, acc) => (acc += item),
+        0
+      );
+      logger.info(`Total cost: ${total_cost}`);
+    }
+  }
+
+  log(event: events.YaEvent): void {
+    // """Register an event."""
+
+    if (this._wrapped_emitter) this._wrapped_emitter(event);
+    if (this.error_occurred) return;
+
+    try {
+      this._handle(event);
+    } catch (error) {
+      console.log("error", error);
+      logger.error("SummaryLogger entered invalid state");
+      this.error_occurred = true;
+    }
+  }
+
+  _handle(event: events.YaEvent) {
+    const eventName = event.constructor.name;
+    if (eventName === events.ComputationStarted.name) this._reset();
+    if (eventName === events.SubscriptionCreated.name)
```

Comment:
Why not use `switch(eventName)` here, instead of a sequence of `if`s?

Reply:
That's right, it can be used here. but I was actually thinking about "lookup table". much performant and may decrease the code line and keep messages and functions organized in an object. if/else was just fastest option without thinking much :)